### PR TITLE
App2App app-native authenticator dummy PR

### PR DESCRIPTION
**Pull request by ImalshaD**
## App2App app-native authenticator dummy PR
App2App app-native authenticator dummy PR
PR to go through the code before the code review.

Issue link: N/A

Doc Issue: N/A

Applicable Labels: N/A

------

### Development Checklist

1. [ ] Build complete solution with pull request in place.
2. [ ] Ran checkstyle plugin with pull request in place.
3. [ ] Ran Findbugs plugin with pull request in place.
4. [ ] Ran FindSecurityBugs plugin and verified report.
5. [ ] Formatted code according to WSO2 code style.
6. [ ] Have you verified the PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [ ] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Verified tests in multiple database environments (if applicable).
3. [ ] Tested with BI enabled  (if applicable).

### CODE REVIEW

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 1]
```diff
@@ -28,23 +30,38 @@
...
 import com.nimbusds.jwt.JWTParser;
 import com.nimbusds.jwt.SignedJWT;
 import com.nimbusds.jwt.proc.ConfigurableJWTProcessor;
 import com.nimbusds.jwt.proc.DefaultJWTProcessor;
 import com.wso2.openbanking.accelerator.common.config.OpenBankingConfigParser;
 import net.minidev.json.JSONObject;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:42]: Remove the repetitive newlines

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 2]
```diff
@@ -120,4 +137,120 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
         return true;
     }
 
+    /**
+     *Validates the signature of a given JWT against a given public key.
+     *
+     * @param signedJWT the signed JWT to be validated
+     * @param publicKey the public key that ois used for validation
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: typo

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 3]
```diff
@@ -120,4 +137,120 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+        }
+        if (StringUtils.isBlank(jwtString)) {
+            return false;
+        }
+        if (StringUtils.countMatches(jwtString, DOT_SEPARATOR) != 2) {
+            return false;
+        }
+        try {
+            //Checking whether the jwtString is jwt parsable.
+            JWTParser.parse(jwtString);
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: We are parsing the jwt in the getSignedJWT method as well. Do we need to parse 2 times?

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 4]
```diff
@@ -120,4 +137,120 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+     */
+    public static SignedJWT getSignedJWT(String jwtString) throws ParseException {
+
+        if (isJWT(jwtString)) {
+            return SignedJWT.parse(jwtString);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Provided token identifier is not a parsable JWT.");
+            }
+            throw new IllegalArgumentException("Provided token identifier is not a parsable JWT.");
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:197]: Better to throw a custom exception we have defined.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 5]
```diff
@@ -120,4 +137,120 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+
+    }
+
+    /**
+     * Validates whether a given JWT is not expired.
+     *
+     * @param signedJWT  the signed jwt that needs to validated
+     * @return true if the jwt is not expired
+     */
+    public static boolean validateExpiryTime(SignedJWT signedJWT) throws ParseException {
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Check if timezone can affect this logic

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/pom.xml [hunk 6]
```diff
@@ -164,13 +164,25 @@
...
+            <artifactId>org.wso2.carbon.identity.application.authenticator.push.common</artifactId>
+        </dependency>
     </dependencies>
 
     <build>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-compiler-plugin</artifactId>
+                <artifactId>maven-compiler-plugin</artifactId><configuration><source>8</source><target>8</target></configuration>
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/pom.xml]: remove configurations

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/model/AppAuthValidationJWT.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import com.wso2.openbanking.accelerator.identity.app2app.validations.annotations.ValidateSignature;
+import com.wso2.openbanking.accelerator.identity.app2app.validations.annotations.ValidateTimeliness;
+import com.wso2.openbanking.accelerator.identity.app2app.validations.validationgroups.RequiredParamChecks;
+import com.wso2.openbanking.accelerator.identity.app2app.validations.validationgroups.ValidityChecks;
+
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+import java.text.ParseException;
+import java.util.Date;
+
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/model/AppAuthValidationJWT.java:33]: remove repetitive newlines

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/utils/App2AppAuthUtils.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            //If matches return the device
+            if (StringUtils.equals(device.getDeviceId(), deviceID)) {
+                String publicKey = deviceHandler.getPublicKey(deviceID);
+                device.setPublicKey(publicKey);
+                return device;
+            }
+        }
+
+        //If no device registered for user matches specified deviceID throw exception
+        throw new IllegalArgumentException("Provided device Identifier does not exist.");
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/utils/App2AppAuthUtils.java]: Use OpenBankingException when throwing

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JTIValidator.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        //adding to cache to prevent the value from being replayed again
+        addToCache(jtiCacheKey, jti);
+        return true;
+
+    }
+
+    private Object getFromCache(IdentityCacheKey identityCacheKey) {
+        //Lazy initialization
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JTIValidator.java]: No need to add comments if the code is self explanatory.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import java.text.ParseException;
+
+/**
+ * Validator class for validating the timeliness of a JWT.
+ * Validates the expiry.
+ * Validates the nbf.
+ */
+public class JwtTokenTimelinessValidator implements ConstraintValidator<ValidateTimeliness, AppAuthValidationJWT> {
+    private static final Log log = LogFactory.getLog(JwtTokenTimelinessValidator.class);
+    @Override
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java]: Missing newline before this

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    public boolean isValid(AppAuthValidationJWT appAuthValidationJWT, ConstraintValidatorContext constraintValidatorContext) {
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java]: Add javadoc comments for public methods. If it is explained in the parent class, use {@inheritdoc}. Check all places.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                return false;
+            }
+
+            //Validating the nbf of the JWT
+            if (!JWTUtils.validateNotValidBefore(signedJWT)) {
+                log.error("JWT is not active.");
+                return false;
+            }
+
+        //Unreachable catch block
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/JwtTokenTimelinessValidator.java]: ?

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/annotations/ValidateJTI.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+package com.wso2.openbanking.accelerator.identity.app2app.validations.annotations;
+
+import com.wso2.openbanking.accelerator.identity.app2app.validations.JTIValidator;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import javax.validation.Constraint;
+import javax.validation.Payload;
+
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/validations/annotations/ValidateJTI.java:29]: Repetitive newlines. Check all places.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/test/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticatorTest.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
+package com.wso2.openbanking.accelerator.identity.app2app;
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/test/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticatorTest.java:19]: Add license header. Check all places

File: open-banking-accelerator/internal-apis/internal-webapps/com.wso2.openbanking.authentication.webapp/src/main/java/com/wso2/openbanking/accelerator/authentication/webapp/OBConsentServlet.java [hunk 15]
```diff
@@ -174,6 +185,8 @@ public void doGet(HttpServletRequest originalRequest, HttpServletResponse respon
 
     }
 
+
```
Comment (imesh94) [inline open-banking-accelerator/internal-apis/internal-webapps/com.wso2.openbanking.authentication.webapp/src/main/java/com/wso2/openbanking/accelerator/authentication/webapp/OBConsentServlet.java:180]: Unnecessary newlines

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
+/**
```
Comment (imesh94) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java:1]: Check if the license header is correct for accelerator.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 5 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Here I am using logic directly implemented in IS 6.1.0 @PushJWTValidator but can't use it as it's private

Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Using epoch time so will not affect.

**Pull request closed without merging**